src/main/java/org/deepsymmetry/beatlink/VirtualRekordbox.java

```diff
@@ -280,13 +286,98 @@ SlotReference findMatchedTrackSourceSlotForPlayer(int player) {
         return playerTrackSourceSlots.get(player);
     }
 
+    /**
+     * Given a player number (normalized to the range 1-4), returns the DeviceSQL rekordbox ID for the
+     * track that is currently loaded on that player. Returns 0 if we are searching for a PSSI match still on a new track load,
+     * or if a PSSI match wasn't found, or if the player does not have a track loaded.
+     *
+     * @param player the player whose DeviceSQL rekordbox ID we are interested in
+     * @return the DeviceSQL rekordbox ID for the track that is currently loaded on that player
+     */
+    int findDeviceSqlRekordboxIdForPlayer(int player) {
+        return playerToDeviceSqlRekordboxId.getOrDefault(player, 0);
+    }
 
     /**
      * Keeps track of the most recent valid (non-zero) status flag byte we have received from each device number,
      * so we can reuse it in cases where a corrupt (zero) value has been sent to us.
      */
     private final Map<Integer, Byte> lastValidStatusFlagBytes = new ConcurrentHashMap<>();
 
+    /**
+     * Tracks packets received from devices to reconstruct complete messages that span multiple packets.
+     * Used primarily for reconstructing binary data (specifically PSSI) from the Opus Quad, which 
+     * arrives fragmented across multiple packets. Maintains state between packet
+     * arrivals until a complete message is assembled, at which point the data can be processed.
+     */
+    private static class PacketTracker {
+        private final List<Byte> data;
+
+        PacketTracker() {
+            this.data = new ArrayList<>();
+        }
+
+        /**
+         * Receives and processes a new packet in the sequence.
+         *
+         * @param packetData the data from the new packet
+         * @param packetNumber which packet this is in the sequence
+         * @param totalPackets how many packets make up the complete message
+         * @return true if this was the final packet in the sequence
+         */
+        boolean receivePacket(byte[] packetData, int packetNumber, int totalPackets) {
+            for (byte b : packetData) {
```

Comment:
I did a bit of poking through the `ArrayList` source code to confirm this, and it would be more efficient to call `addAll()` with the entirety of the new `packetData`, since that will resize the array at most once, rather than iterating a byte at a time and repeatedly calling `add()`..

Reply:
As we discussed on Zulip, once actually trying this approach it turns out I had forgotten limitations of auto-boxing with respect to primitive arrays, and this actually won’t work. The approach you took is in fact the best way to do it, short of using `ByteBuffer` instead of an array list of bytes. I had thought about suggesting that in my first review pass, but decided against it, because even though it is more efficient and the “right” (idiomatic) way to work with clumps of bytes like this, and I use it other places in Beat Link, it would be a fairly complicated API for you to learn and adopt. So let’s stick with this approach for now, and maybe someday one of us will get ambitious and rewrite this to use a `ByteBuffer` instead.